### PR TITLE
Feat/file graph ql

### PR DESCRIPTION
**Pull request by claudia1296**
# FEATUER 

closes #Clin-16 et clin-196

## Description

Page fichier avec GraphQl et Apollo

## Validation

- [ ] Test Coverage
- [ ] QA Done
- [ ] Design/UI Approved from design

## QA

Étapes de validation
Url (storybook, ...)
...

## Notification

@ QA, Design ...

### CODE REVIEW

File: client/src/store/graphql/files/actions.tsx [hunk 1]
```diff
@@ -0,0 +1,77 @@
+import { useLazyResultQuery } from 'store/graphql/utils/query';
+import { gql } from '@apollo/client';
+
+export type QueryVariable = {
+  patientId: string;
+};
+
+export const getFilesData = (variables: QueryVariable) => () => {
+  const test = gql`{
```
Comment (adipaul1981) [inline client/src/store/graphql/files/actions.tsx:9]: Maybe call this constant 'query'. Makes more sense that 'test'.

File: client/src/store/graphql/files/queries.tsx [hunk 2]
```diff
@@ -0,0 +1,64 @@
+import { gql } from '@apollo/client';
+
+export const FILES_QUERY = gql`
+{
+  Patient(id:"QA-PA-00128"){
```
Comment (adipaul1981) [inline client/src/store/graphql/files/queries.tsx:5]: I'm confused here. is this for a test? Why there is a hardcoded patient id in the query? is this query used at all?

File: client/src/components/screens/Patient/components/FilesTab/index.tsx [hunk 3]
```diff
@@ -4,41 +4,37 @@ import {
...
-      return `${newSize} ${intl.get('screen.patient.details.file.size.mb')}`;
-    } if (size >= 10 ** 9 && size < 10 ** 12) {
-      newSize = size / 10 ** 9;
-      return `${newSize} ${intl.get('screen.patient.details.file.size.gb')}`;
-    }
-
-    return `${newSize} ${intl.get('screen.patient.details.file.size.b')}`;
-  };
+  const { loading: loadingData, results: data } = getFilesData({ patientId: patient.id })();
+  if (loadingData) {
```
Comment (adipaul1981) [inline client/src/components/screens/Patient/components/FilesTab/index.tsx:35]: In the Apollo query, you return the previous data is the current data is not available yet. Here you check if its loading and if so return null (regardless of anything else).

Maybe check what is the desired behavior of the portal in this situation. 
- If you have some data (previous) and is loading, do you still what to show the component with it until the new data comes in?

- or maybe, if loading and you have previous data, you just want to show an empty card, until the data comes in? (in that case maybe add a spinner?)

- or maybe this is the desired behavior (return null if loading) and then disregard this comment :)

File: client/src/components/screens/Patient/components/FilesTab/components/EmptyCard.tsx [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+import IconKit from 'react-icons-kit';
+import { useSelector, useDispatch } from 'react-redux';
+import {
+  ic_cloud_download,
+} from 'react-icons-kit/md';
+import intl from 'react-intl-universal';
+import { State } from 'reducers';
+import { navigateToPatientScreen } from 'actions/router';
+
+const EmptyCard: React.FC = () => {
```
Comment (evans-g-crsj) [inline client/src/components/screens/Patient/components/FilesTab/components/EmptyCard.tsx:14]: remove React.FC , pls

File: client/src/components/screens/Patient/components/FilesTab/index.tsx [hunk 5]
```diff
@@ -4,41 +4,37 @@ import {
...
   return data.payload.data.url;
 };
 
-const FilesTab: React.FC = () => {
-  const { Patient } = fileInfo;
+type FilesTabProps = {
+  userToken: string;
+};
 
+const FilesTab: React.FC<FilesTabProps> = () => {
```
Comment (evans-g-crsj) [inline client/src/components/screens/Patient/components/FilesTab/index.tsx:29]: no React.Fc<stuff>. Plus the FilesTabProps are not used anywhere.

File: client/src/components/screens/Patient/components/FilesTab/index.tsx [hunk 6]
```diff
@@ -51,7 +47,7 @@ const FilesTab: React.FC = () => {
     });
   };
 
-  const getDropdownOption = (format: string, url: any, documentR: string) => {
+  const getDropdownOption = (format: string, url: any) => {
```
Comment (evans-g-crsj) [inline client/src/components/screens/Patient/components/FilesTab/index.tsx:50]: if possible, find a type vs any

File: client/src/components/screens/Patient/components/FilesTab/index.tsx [hunk 7]
```diff
@@ -4,41 +4,37 @@ import {
...
 import MetadataModal from './metadataModal';
+import { State } from '../../../../../reducers';
+import { ParsedPatientData } from '../../../../../helpers/providers/types';
+import EmptyCard from './components/EmptyCard';
 
 import './styles.scss';
 
-const fileInfo: {[key: string]: any} = require('./info.json');
-
 const getURL = async (url: string) => {
```
Comment (evans-g-crsj) [inline client/src/components/screens/Patient/components/FilesTab/index.tsx:20]: should be in redux

File: client/src/components/screens/Patient/components/FilesTab/index.tsx [hunk 8]
```diff
@@ -122,106 +118,59 @@ const FilesTab: React.FC = () => {
...
-            },
-            {
-              key: 'action',
-              dataIndex: 'action',
-              title: intl.get('screen.patient.details.file.action'),
-            },
-          ]}
-          dataSource={dataSource}
-          size="small"
+  if (data.docs) {
```
Comment (evans-g-crsj) [inline client/src/components/screens/Patient/components/FilesTab/index.tsx:121]: if (!data.docs) { return <EmptyCard/> }

File: client/src/components/screens/Patient/components/FilesTab/index.tsx [hunk 9]
```diff
@@ -122,106 +118,59 @@ const FilesTab: React.FC = () => {
...
+    data.docs.forEach((element: PatientResponse) => {
+      const {
```
Comment (evans-g-crsj) [inline client/src/components/screens/Patient/components/FilesTab/index.tsx:123]: I think this logic should be extracted.

File: client/src/components/screens/Patient/index.jsx [hunk 10]
```diff
@@ -302,7 +302,7 @@ class PatientScreen extends React.Component {
             Fichiers
           </span>
         ),
-        content: <FilesTab />,
+        content: <ApolloProvider userToken={keycloak.token}><FilesTab /></ApolloProvider>,
```
Comment (evans-g-crsj) [inline client/src/components/screens/Patient/index.jsx:305]: pls, add a // FIXME and create a ticket for this.

File: client/src/store/graphql/files/actions.tsx (hunk 1 above)
Comment (evans-g-crsj) [reply client/src/store/graphql/files/actions.tsx:9]: Agreed. Moreover, maybe it should be in a file of it's own. but if it s the only graphql query so far you can let it here for now.

File: client/src/store/graphql/files/models.tsx [hunk 11]
```diff
@@ -0,0 +1,60 @@
+import intl from 'react-intl-universal';
```
Comment (evans-g-crsj) [inline client/src/store/graphql/files/models.tsx:1]: intl can only be used in react components. What you did with the intKey is excellent. Do not use intl here. Plus, if it is a UI column than do not put it in this file. It must be in a react component.

**Review by evans-g-crsj — CHANGES_REQUESTED** [review]

File: client/src/components/screens/Patient/components/FilesTab/index.tsx [hunk 12]
```diff
@@ -1,67 +1,116 @@
...
-
-    return `${newSize} ${intl.get('screen.patient.details.file.size.b')}`;
+  const { loading: loadingData, results: data } = getFilesData({ patientId: patient.id })();
+  if (loadingData) {
+    return null;
+  }
+
+  const getURL = (url: string) => {
+    const dataUrl = dispatch(getPatientFileURL(url));
+    return dataUrl.payload.file;
```
Comment (evans-g-crsj) [inline client/src/components/screens/Patient/components/FilesTab/index.tsx:35]: tu y es presque, mais c'est pas tout-à-fait ça. Il faut que tu `dispatch` et que dans ton store des fichiers tu récupères l'url du file avec un sélecteur.

File: client/src/utils.tsx [hunk 13]
```diff
@@ -0,0 +1,16 @@
+export const formatBytes = (bytes: number, decimals = 2) => {
```
Comment (evans-g-crsj) [inline client/src/utils.tsx:1]: a file called utils is too general. As a slight improvement create a file called fileHelper in the directory where there are other similar helpers.

**Review by evans-g-crsj — CHANGES_REQUESTED** [review]

File: client/src/components/screens/Patient/components/FilesTab/index.tsx [hunk 14]
```diff
@@ -1,67 +1,116 @@
...
-import Api from '../../../../../helpers/api';
-import { TaskResponse } from '../../../../../helpers/search/types';
-import MetadataModal from './metadataModal';
+import { useDispatch,useSelector } from 'react-redux';
+import { DownOutlined } from '@ant-design/icons';
+import { getPatientFileURL } from 'actions/patient';
+import {
+Button,   Card, Dropdown, Menu,
+Table, } from 'antd';
+import { formatBytes } from 'helpers/file';
```
Comment (francisl) [inline client/src/components/screens/Patient/components/FilesTab/index.tsx:9]: Le linter devrait faire une erreur ici.

File: client/src/components/screens/Patient/components/FilesTab/index.tsx [hunk 15]
```diff
@@ -123,105 +172,63 @@ const FilesTab: React.FC = () => {
...
+      };
+
+      dataSource.push(dataLine);
+    });
+  };
+
+  if (data.docs) {
+    getData();
+    return (
+      <div className="page-static-content files-tab">
```
Comment (francisl) [inline client/src/components/screens/Patient/components/FilesTab/index.tsx:212]: Pour la simplicité de lecture du code, j'inverserait la condition. Retourne le EmptyCard immédiatement. Comme ca tu va sauver une indentation et un embranchement lourd.

File: client/src/components/screens/Patient/index.jsx [hunk 16]
```diff
@@ -16,7 +15,9 @@ import {
   navigateToPatientVariantScreen,
   navigateToSubmissionWithPatient,
 } from '../../../actions/router';
+import keycloak from '../../../keycloak';
 import { appShape } from '../../../reducers/app';
+import ApolloProvider from '../../../store/providers/apollo';
 import Layout from '../../Layout';
 import PatientVariantScreen from '../PatientVariant';
```
Comment (francisl) [inline client/src/components/screens/Patient/index.jsx:22]: svp path absolue

File: client/src/helpers/file.tsx [hunk 17]
```diff
@@ -0,0 +1,17 @@
+export const formatBytes = (bytes: number, decimals = 2) => {
+    if (bytes === 0) {
+      return '0';
+    }
+  
+    const scale = 1024;
```
Comment (francisl) [inline client/src/helpers/file.tsx:6]: cette constant peut être placé a un plus haut niveau. au lieu de prendre de la mémore a chaque appel de fonction. 
Même chose pour sizes. 

Petite optimisation qui n'ajoute aucune complexité au code.

File: client/src/store/graphql/utils/query.ts [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+    data, error, loading, previousData,
+  } = useQuery<TData, TVariables>(query, options);
+
+  let result = previousData;
+
+  if (data) {
+    result = data;
+  }
+
+  return { error, loading, result };
```
Comment (francisl) [inline client/src/store/graphql/utils/query.ts:36]:
```
const result = data ? data : previousData;
return { error, loading, result };
```

**Review by francisl — CHANGES_REQUESTED** [review]

File: client/src/components/screens/Patient/components/FilesTab/index.tsx (hunk 14 above)
Comment (claudia1296) [reply client/src/components/screens/Patient/components/FilesTab/index.tsx:9]: ici je ne suis pas capable d'avoir l'erreur de linter

Comment (francisl) [reply client/src/components/screens/Patient/components/FilesTab/index.tsx:9]: peut etre la regle ne l'attrape pas
enter avant `}`

dans ce cas, laisse le comme ca on va fixé le linter plus tard

**Review by francisl — DISMISSED** [review]

File: client/src/sagas/patient.js [hunk 19]
```diff
@@ -341,6 +341,7 @@ function* getFileURL(action) {
   try {
     const { file } = action.payload;
     const fileURL = yield Api.getFileURL(file);
+    window.open(fileURL.payload.data.url , '_blank');
     if (fileURL.error) {
```
Comment (evans-g-crsj) [inline client/src/sagas/patient.js:345]: error before anything else.

**Review by evans-g-crsj — APPROVED** [review]

**Review by evans-g-crsj — APPROVED** [review]

**Pull request merged**